### DISCUSSION TITLE
bash printf, coreutils printf, and numfmt not working as expected comparted to all Linux distros

### DISCUSSION THREAD

**Discussion by theofficialgman** [community/community#55358] [Actions] [answered]
### Select Topic Area

Bug

### Body

I expect the following to all produce the output `123,456,789`

however the following commands produce incorrect output `123456789` on github actions
coreutils printf:
```
/usr/bin/printf "%'d" 123456789
```
bash builtin printf
```
printf "%'d" 123456789
```
coreutils numfmt
```
numfmt --grouping 123456789
```

Tested on multiple linux distros on actual hardware (ubuntu 20.04/22.04/23.04 and piOS bullseye) and all of the above produce `123,456,789`

This appears to be a bug though I am not sure how actions runner is causing it.

**Comment by airtower-luna** [comment]
I'm curious, exactly how do the steps look in the workflow file where you try to format numbers?

**Reply by theofficialgman** [reply]
it does not matter how they are called. I can put them in a file (eg, test.sh) and call that directly with a bash shebang at the top of the file, or put them directly after run

https://github.com/theofficialgman/testing/blob/9f6fe9737f2465f3bf00ecc76e79d7a35ff87824/.github/workflows/test_CI.yml

**Reply by ChristopherHX** [reply]
You have configured your locale locally, while my `ubuntu 20.04` outputs only `123456789` for all your examples using `C.UTF8` locale. Eventually is GitHub Actions using the same default locale.

> ‘--grouping’
Group digits in output numbers according to the current locale’s grouping rules (e.g Thousands Separator character, commonly ‘.’ (dot) or ‘,’ comma). This option has no effect in ‘POSIX/C’ locale.

https://www.gnu.org/software/coreutils/manual/html_node/numfmt-invocation.html#numfmt-invocation

**Reply by theofficialgman** [reply]
I see. Unfortunatly it can't be fully changed within the context of github actions (shell builtin functions don't respect locale changes even if you start a new job)

https://github.com/actions/runner-images/issues/762#issuecomment-618415592

**Comment by ChristopherHX** [comment, marked as answer]
You can set your locale to en-US. First example is for job container running with user root and don't have sudo installed.
```yaml
on: push
jobs:
  _:
    runs-on: ubuntu-latest
    container: node:latest
    steps:
    - name: SetLocale
      shell: bash
      run: |
              apt-get update && apt-get install locales-all -y
              echo "LANG=en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_CTYPE="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_NUMERIC="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_TIME="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_COLLATE="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_MONETARY="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_MESSAGES="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_PAPER="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_NAME="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_ADDRESS="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_TELEPHONE="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_MEASUREMENT="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_IDENTIFICATION="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_ALL="en_US.UTF-8" >> "$GITHUB_ENV"     
    - name: Test
      shell: bash
      run: |
              /usr/bin/printf "%'d" 123456789
              printf "%'d" 123456789
              numfmt --grouping 123456789
```

For GitHub Hosted runners (no container) this might work
```yaml
on: push
jobs:
  _:
    runs-on: ubuntu-latest
    steps:
    - name: SetLocale
      shell: bash
      run: |
              sudo apt-get update && sudo apt-get install locales-all -y
              echo "LANG=en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_CTYPE="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_NUMERIC="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_TIME="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_COLLATE="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_MONETARY="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_MESSAGES="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_PAPER="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_NAME="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_ADDRESS="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_TELEPHONE="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_MEASUREMENT="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_IDENTIFICATION="en_US.UTF-8" >> "$GITHUB_ENV"
              echo LC_ALL="en_US.UTF-8" >> "$GITHUB_ENV"     
    - name: Test
      shell: bash
      run: |
              /usr/bin/printf "%'d" 123456789
              printf "%'d" 123456789
              numfmt --grouping 123456789
```
You can also rewrite the `GITHUB_ENV` steps to the `jobs.<id>.env` as a yml mapping

**Reply by theofficialgman** [reply]
I can confirm that the "For GitHub Hosted runners (no container) this might work" method does work for all three commands.

I'll mark this as the answer and close this
